api/v1alpha1/warehouse_types.go

```diff
@@ -0,0 +1,171 @@
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// +kubebuilder:validation:Enum={SemVer,NewestBuild,Alphabetical,Digest}
+type ImageUpdateStrategy string
+
+const (
+	ImageUpdateStrategySemVer       ImageUpdateStrategy = "SemVer"
+	ImageUpdateStrategyNewestBuild  ImageUpdateStrategy = "NewestBuild"
+	ImageUpdateStrategyAlphabetical ImageUpdateStrategy = "Alphabetical"
+	ImageUpdateStrategyDigest       ImageUpdateStrategy = "Digest"
+)
+
+//+kubebuilder:object:root=true
+//+kubebuilder:subresource:status
+
+// Warehouse is a source of Freight.
+type Warehouse struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// Spec describes sources of artifacts.
+	//
+	//+kubebuilder:validation:Required
+	Spec *WarehouseSpec `json:"spec"`
+	// Status describes the Warehouse's most recently observed state.
+	Status WarehouseStatus `json:"status,omitempty"`
+}
+
+func (w *Warehouse) GetStatus() *WarehouseStatus {
+	return &w.Status
+}
+
+// WarehouseSpec describes sources of versioned artifacts to be included in
+// Freight produced by this Warehouse.
+type WarehouseSpec struct {
+	// Subscriptions describes sources of artifacts to be included in Freight
+	// produced by this Warehouse.
+	//
+	//+kubebuilder:validation:Required
+	Subscriptions *RepoSubscriptions `json:"subscriptions"`
+}
+
+// RepoSubscriptions describes various sorts of repositories a Warehouse uses
+// as sources of versioned artifacts.
+type RepoSubscriptions struct {
+	// Git describes subscriptions to Git repositories.
+	Git []GitSubscription `json:"git,omitempty"`
+	// Images describes subscriptions to container image repositories.
+	Images []ImageSubscription `json:"images,omitempty"`
+	// Charts describes subscriptions to Helm charts.
+	Charts []ChartSubscription `json:"charts,omitempty"`
+}
+
+// GitSubscription defines a subscription to a Git repository.
+type GitSubscription struct {
+	// URL is the repository's URL. This is a required field.
+	//
+	//+kubebuilder:validation:MinLength=1
+	//+kubebuilder:validation:Pattern=`^https://(\w+([\.-]\w+)*@)?\w+([\.-]\w+)*(:[\d]+)?(/.*)?$`
+	RepoURL string `json:"repoURL"`
+	// Branch references a particular branch of the repository. This field is
+	// optional. When not specified, the subscription is implicitly to the
+	// repository's default branch.
+	//
+	//+kubebuilder:validation:MinLength=1
+	//+kubebuilder:validation:Pattern=`^\w+([-/]\w+)*$`
+	Branch string `json:"branch,omitempty"`
+}
+
+// ImageSubscription defines a subscription to an image repository.
+type ImageSubscription struct {
+	// RepoURL specifies the URL of the image repository to subscribe to. The
+	// value in this field MUST NOT include an image tag. This field is required.
+	//
+	//+kubebuilder:validation:MinLength=1
+	//+kubebuilder:validation:Pattern=`^(\w+([\.-]\w+)*(:[\d]+)?/)?(\w+([\.-]\w+)*)(/\w+([\.-]\w+)*)*$`
```

Comment:
We need to make we don't regress https://github.com/akuity/kargo/pull/915

Reply:
Double-check me, but I think I copied this from #915 this morning.

Reply:
:embarrassed: https://github.com/akuity/kargo/pull/931

Reply:
No worries. I'm afk for a few days. Will look soon.